### PR TITLE
Feature/#358 update edc 0.7.0

### PR DESCRIPTION
**Pull request by ds-jhartmann**
story: https://github.com/eclipse-tractusx/item-relationship-service/issues/358

### CODE REVIEW

File: charts/item-relationship-service/CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - Fix for incomplete renaming of Helm chart (see #489)
-
+- Fixed missing Spring attribute
```
Comment (ds-mwesener) [inline charts/item-relationship-service/CHANGELOG.md:12]: Ticket number if existing.

File: irs-api/src/test/java/org/eclipse/tractusx/irs/IrsWireMockIntegrationTest.java [hunk 2]
```diff
@@ -280,8 +278,7 @@ private void successfulRegistryAndDataRequest(final String globalAssetId, final
             final String batchFileName, final String sbomFileName) {
 
         final String edcAssetId = WiremockSupport.randomUUIDwithPrefix();
-        final String batch = WiremockSupport.submodelRequest(edcAssetId, "Batch",
-                batchAspectName, batchFileName);
+        final String batch = WiremockSupport.submodelRequest(edcAssetId, "Batch", batchAspectName, batchFileName);
```
Comment (ds-mwesener) [inline irs-api/src/test/java/org/eclipse/tractusx/irs/IrsWireMockIntegrationTest.java:281]: Why not using constants for the type? "Batch"

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 3]
```diff
@@ -88,30 +97,26 @@ public List<CatalogItem> fetchCatalogItems(final CatalogRequest catalogRequest)
...
-                                                   .connectorId(catalog.getId())
-                                                   .offerId(offer.getKey())
-                                                   .policy(offer.getValue());
-                if (catalog.getProperties().containsKey(JsonLdConfiguration.NAMESPACE_EDC_PARTICIPANT_ID)) {
-                    catalogItem.connectorId(
-                            catalog.getProperties().get(JsonLdConfiguration.NAMESPACE_EDC_PARTICIPANT_ID).toString());
-                }
-
-                return catalogItem.build();
+            return catalog.getDatasets().stream().map(dataset -> {
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:105]: use emptyIfNull or similiar

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 4]
```diff
@@ -152,33 +159,32 @@ public List<CatalogItem> fetchCatalogItemsUntilMatch(final String connectorUrl,
...
+     * @return The list of catalog Items matching the target id
+     */
+    @Deprecated(since = "5.0.0")
+    public List<CatalogItem> fetchCatalogById(final String connectorUrl, final String target, final String bpn) {
+        return fetchCatalogByFilter(connectorUrl, NAMESPACE_EDC_ID, target, bpn);
     }
 
-    public List<CatalogItem> fetchCatalogByFilter(final String connectorUrl, final String key, final String value) {
-        final Catalog catalog = controlPlaneClient.getCatalogWithFilter(connectorUrl, key, value);
+    @SuppressWarnings("PMD.UseObjectForClearerAPI") // TODO (ds-jhartmann) create object
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:174]: TODO - leave it if you want to adress it later.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 5]
```diff
@@ -152,33 +159,32 @@ public List<CatalogItem> fetchCatalogItemsUntilMatch(final String connectorUrl,
...
+    /**
+     * @deprecated
+     * @param connectorUrl The EDC Connector from which the Catalog will be requested
+     * @param target       The target assetID which will be searched for
+     * @param bpn          The BPN of the company to which the EDC Connector belongs
+     * @return The list of catalog Items matching the target id
+     */
+    @Deprecated(since = "5.0.0")
+    public List<CatalogItem> fetchCatalogById(final String connectorUrl, final String target, final String bpn) {
+        return fetchCatalogByFilter(connectorUrl, NAMESPACE_EDC_ID, target, bpn);
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:171]: Check if null is possible.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 6]
```diff
@@ -152,33 +159,32 @@ public List<CatalogItem> fetchCatalogItemsUntilMatch(final String connectorUrl,
...
+        final Catalog catalog = controlPlaneClient.getCatalogWithFilter(connectorUrl, key, value, bpn);
         return mapToCatalogItems(catalog);
     }
 
     private Optional<Dataset> findOfferIfExist(final String target, final Catalog catalog) {
-        return catalog.getDatasets()
-                      .stream()
-                      .filter(dataset -> dataset.getProperty(NAMESPACE_EDC_ID).toString().equals(target))
-                      .findFirst();
+        return catalog.getDatasets().stream().filter(dataset -> dataset.getId().equals(target)).findFirst();
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:182]: null?

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 7]
```diff
@@ -152,33 +159,32 @@ public List<CatalogItem> fetchCatalogItemsUntilMatch(final String connectorUrl,
...
-                                                          .map(dataset -> dataset.getProperty(NAMESPACE_EDC_ID)
-                                                                                 .toString())
-                                                          .collect(toSet());
-        final Set<String> nextOffers = newPageableCatalog.getDatasets()
-                                                         .stream()
-                                                         .map(dataset -> dataset.getProperty(NAMESPACE_EDC_ID)
-                                                                                .toString())
-                                                         .collect(toSet());
+        final Set<String> previousOffers = pageableCatalog.getDatasets().stream().map(Dataset::getId).collect(toSet());
+        final Set<String> nextOffers = newPageableCatalog.getDatasets().stream().map(Dataset::getId).collect(toSet());
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:187]: null?

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/model/NegotiationRequest.java [hunk 8]
```diff
@@ -51,16 +52,14 @@ public class NegotiationRequest {
     public static final String NEGOTIATION_ASSET_ID = "https://w3id.org/edc/v0.0.1/ns/assetId";
     public static final String NEGOTIATION_POLICY = "https://w3id.org/edc/v0.0.1/ns/policy";
 
-    @NotBlank(message = "connectorAddress is mandatory")
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/model/NegotiationRequest.java:54]: Add the private fields

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java [hunk 9]
```diff
@@ -84,37 +84,40 @@ public class EdcTransformer {
...
-    private final org.eclipse.tractusx.irs.edc.client.transformer.JsonObjectToPolicyTransformer jsonObjectToPolicyTransformer;
     private final TitaniumJsonLd titaniumJsonLd;
     private final TransformerContextImpl transformerContext;
     private final JsonObjectFromAssetTransformer jsonObjectFromAssetTransformer;
+    private final JsonObjectToIrsPolicyTransformer jsonObjectToIrsPolicyTransformer;
 
     public EdcTransformer(@Qualifier(JSON_LD_OBJECT_MAPPER) final ObjectMapper objectMapper,
             final TitaniumJsonLd titaniumJsonLd, final TypeTransformerRegistry typeTransformerRegistry) {
         this.titaniumJsonLd = titaniumJsonLd;
+        transformerContext = new TransformerContextImpl(typeTransformerRegistry);
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java:95]: Let spring manage the instance

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/JsonObjectToDatasetTransformer.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            final var policy = context.transform(object, Policy.class);
+            builder.offer(offerId, policy);
+        } else if (value instanceof JsonArray array) {
+            array.forEach(entry -> transformPolicies(entry, builder, context));
+        } else {
+            context.problem()
+                   .unexpectedType()
+                   .type(DCAT_DATASET_TYPE)
+                   .property(ODRL_POLICY_ATTRIBUTE)
+                   .actual(value == null ? "null" : value.getValueType().toString())
```
Comment (ds-mwesener) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/JsonObjectToDatasetTransformer.java:79]: Is this necessary?

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java [hunk 11]
```diff
@@ -88,30 +97,26 @@ public List<CatalogItem> fetchCatalogItems(final CatalogRequest catalogRequest)
...
-                            catalog.getProperties().get(JsonLdConfiguration.NAMESPACE_EDC_PARTICIPANT_ID).toString());
-                }
-
-                return catalogItem.build();
+            return catalog.getDatasets().stream().map(dataset -> {
+                final Map.Entry<String, Policy> offer = dataset.getOffers()
+                                                               .entrySet()
+                                                               .stream()
+                                                               .findFirst()
+                                                               .orElseThrow();
```
Comment (dsmf) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:110]: maybe throw an exception with a message that explains what went wrong from a business point of view

**Review by ds-mwesener — CHANGES_REQUESTED** [review]
Please see comments

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java (hunk 4 above)
Comment (dsmf) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:174]: suggestion: remove TODO or link to issue. Could be addressed in context of https://github.com/eclipse-tractusx/item-relationship-service/issues/547.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java (hunk 9 above)
Comment (ds-jhartmann) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java:95]: EdcTransformer is created as bean. There is no bean for TransformerContext by default since this class is added via the EDC dependencies. 
I don't see the benefit of creating a dedicated bean for TransformerContext here since it is just used in this class

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/JsonObjectToDatasetTransformer.java (hunk 10 above)
Comment (ds-jhartmann) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/JsonObjectToDatasetTransformer.java:79]: I reused implementations from previous EDC versions here. I would not like to remove this without thorough testing

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java [hunk 12]
```diff
@@ -48,19 +57,50 @@ public class EdcCallbackController {
...
-        final String authCode = dataReference.getAuthCode();
-        if (authCode != null) {
-            final var contractAgreementId = EDRAuthCode.fromAuthCodeToken(authCode).getCid();
-            storage.put(contractAgreementId, dataReference);
-            log.info("Endpoint Data Reference received and cached for agreement: {}", Masker.mask(contractAgreementId));
-        } else {
-            log.error("ContractAgreementId could not be extracted from Endpoint Data Reference {}",
-                    StringMapper.mapToString(dataReference));
+    public void receiveEdcCallback(final @RequestBody String endpointDataReferenceCallback) {
+        EndpointDataReference endpointDataReference;
```
Comment (dsmf) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java:61]:
```suggestion
        final EndpointDataReference endpointDataReference;
```

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java [hunk 13]
```diff
@@ -48,19 +57,50 @@ public class EdcCallbackController {
...
+        } catch (EdcClientException e) {
+            log.error("Could not deserialize Endpoint Data Reference {}", endpointDataReferenceCallback);
         }
+
+    }
+
+    private static EndpointDataReference mapToEndpointDataReference(final String endpointDataReference)
+            throws EdcClientException {
+        try {
+            EndpointDataReference dataReference;
```
Comment (dsmf) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java:81]:
```suggestion
            final EndpointDataReference dataReference;
```
Furthermore declaration should be outside the try just like it is done in the other method.
Reason: Makes it clear at once that this variable contains the methods results.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java (hunk 4 above)
Comment (ds-jhartmann) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EDCCatalogFacade.java:171]: added link to the issue

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java (hunk 13 above)
Comment (ds-jhartmann) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java]: done

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java (hunk 12 above)
Comment (ds-jhartmann) [reply irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackController.java]: done

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/model/edr/Payload.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+ * @param transferProcessId  The ID of the transfer process.
+ * @param callbackAddresses  The list of callback addresses.
+ * @param assetId            The ID of the EDC asset.
+ * @param type               The type of payload.
+ * @param contractId         The ID of the contract agreement.
+ * @param dataAddress        The data address.
+ */
+@Builder
+@Jacksonized
+public record Payload(String transferProcessId,
```
Comment (dsmf) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/model/edr/Payload.java:39]: Hi @ds-jhartmann , might be good to rename to something less general like TransferProcessPayload.

File: irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java [hunk 15]
```diff
@@ -84,37 +84,40 @@ public class EdcTransformer {
     private final JsonObjectFromTransferProcessRequestTransformer jsonObjectFromTransferProcessRequestTransformer;
     private final JsonObjectFromContractOfferDescriptionTransformer jsonObjectFromContractOfferDescriptionTransformer;
     private final JsonObjectFromCatalogRequestTransformer jsonObjectFromCatalogRequestTransformer;
-    private final org.eclipse.tractusx.irs.edc.client.transformer.JsonObjectToPolicyTransformer jsonObjectToPolicyTransformer;
     private final TitaniumJsonLd titaniumJsonLd;
     private final TransformerContextImpl transformerContext;
```
Comment (dsmf) [inline irs-edc-client/src/main/java/org/eclipse/tractusx/irs/edc/client/transformer/EdcTransformer.java:88]:
```suggestion
    private final TransformerContext transformerContext;
```

File: irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java [hunk 16]
```diff
@@ -38,31 +39,120 @@ class EdcCallbackControllerTest {
...
         // assert
-        final var result = storage.get("testId");
+        final var result = storage.get("testContractId");
         assertThat(result).isNotNull().isEmpty();
     }
+
+    @Test
+    void shouldDeserializeCallbackEDR() {
+        // arrange
+        String callbackEDR = """
```
Comment (dsmf) [inline irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java:105]:
```suggestion
        final String callbackEDR = """
```

File: irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java [hunk 17]
```diff
@@ -38,31 +39,120 @@ class EdcCallbackControllerTest {
...
+        assertThat(result.get().getEndpoint()).isEqualTo(expectedEndpoint);
+        assertThat(result.get().getAuthKey()).isEqualTo(expectedAuthKey);
+        assertThat(result.get().getAuthCode()).isEqualTo(expectedAuthCode);
     }
 
     @Test
     void shouldDoNothingWhenEDRTokenIsInvalid() {
         // arrange
-        final var ref = EndpointDataReference.Builder.newInstance().endpoint("test").build();
+        String ref = """
```
Comment (dsmf) [inline irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java:83]:
```suggestion
        final String ref = """
```

File: irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java [hunk 18]
```diff
@@ -38,31 +39,120 @@ class EdcCallbackControllerTest {
     @Test
     void shouldStoreAgreementId() {
         // arrange
-        final var ref = EndpointDataReference.Builder.newInstance()
-                                                     .endpoint("test")
-                                                     .authKey("Authorization")
-                                                     .authCode(
-                                                             "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2ODkwODA5OTEsImRhZCI6InRlc3QiLCJjaWQiOiJ0ZXN0SWQiLCJpYXQiOjE2ODkwODI2ODF9.62AIg-k8Yz6xLUBPblv2AtA5fuhoBnm9KMxhdCUunhA")
-                                                     .build();
+        String ref = """
```
Comment (dsmf) [inline irs-edc-client/src/test/java/org/eclipse/tractusx/irs/edc/client/EdcCallbackControllerTest.java:42]:
```suggestion
        final String ref = """
```

File: irs-policy-store/pom.xml [hunk 19]
```diff
@@ -85,6 +85,13 @@
             <artifactId>irs-common</artifactId>
             <version>${irs-registry-client.version}</version>
         </dependency>
+        <!-- https://mvnrepository.com/artifact/org.eclipse.jetty.toolchain/jetty-jakarta-servlet-api -->
+        <dependency>
+            <groupId>org.eclipse.jetty.toolchain</groupId>
```
Comment (dsmf) [inline irs-policy-store/pom.xml:90]: Hi @ds-jhartmann, for what is this dependency used?

File: irs-registry-client/src/test/java/org/eclipse/tractusx/irs/registryclient/decentral/DecentralDigitalTwinRegistryClientTest.java [hunk 20]
```diff
@@ -35,9 +35,9 @@ void setUp() {
     @Test
     void shouldCallForAssetAdministrationShellDescriptor() {
         // given
-        EndpointDataReference endpointDataReference = EndpointDataReference.Builder.newInstance()
-                                                                                   .endpoint("url.to.host")
-                                                                                   .build();
+        final String contractAgreementId = "contractAgreementId";
+        final String endpointUrl = "url.to.host";
+        EndpointDataReference endpointDataReference = endpointDataReference(contractAgreementId, endpointUrl);
```
Comment (dsmf) [inline irs-registry-client/src/test/java/org/eclipse/tractusx/irs/registryclient/decentral/DecentralDigitalTwinRegistryClientTest.java:40]:
```suggestion
        final EndpointDataReference endpointDataReference = endpointDataReference(contractAgreementId, endpointUrl);
```

File: pom.xml [hunk 21]
```diff
@@ -92,7 +92,7 @@
         <snakeyaml.version>2.0</snakeyaml.version>
         <commons-validator.version>1.8.0</commons-validator.version>
         <swagger-annotations.version>2.2.18</swagger-annotations.version>
-        <edc.version>0.2.1</edc.version>
+        <edc.version>0.6.0</edc.version>
```
Comment (dsmf) [inline pom.xml:95]: Hi @ds-jhartmann, shoudn't it be 0.7.0?

**Review by dsmf — CHANGES_REQUESTED** [review]
see my comments

File: irs-policy-store/pom.xml (hunk 19 above)
Comment (ds-jhartmann) [reply irs-policy-store/pom.xml:90]: For the parameter validation which you introduced to the PolicyStoreController

File: pom.xml (hunk 21 above)
Comment (ds-jhartmann) [reply pom.xml:95]: No, the latest version of the EDC packages is 0.6.0

File: irs-api/src/test/java/org/eclipse/tractusx/irs/IrsWireMockIntegrationTest.java [hunk 22]
```diff
@@ -111,12 +112,6 @@ static void startContainer() {
         WiremockSupport.successfulSemanticModelRequest();
     }
 
-    @AfterEach
-    void tearDown() {
-        cacheManager.getCacheNames()
-                    .forEach(cacheName -> Objects.requireNonNull(cacheManager.getCache(cacheName)).clear());
```
Comment (dsmf) [inline irs-api/src/test/java/org/eclipse/tractusx/irs/IrsWireMockIntegrationTest.java:117]: Hi @ds-jhartmann, why was this removed?

Comment (ds-jhartmann) [reply irs-api/src/test/java/org/eclipse/tractusx/irs/IrsWireMockIntegrationTest.java:117]: Was not removed, only re-arranged

**Review by dsmf — APPROVED** [review]
LGTM
